src/main.rs

```diff
@@ -61,69 +59,122 @@ impl BFInterp {
         }
     }
 
-    fn it_bb(&mut self, bb: &BasicBlock) -> Option<u64> {
-        for inst in bb.instrs.iter() {
-            match inst {
-                Instruction::Alloca(_) => {}
-                Instruction::Store(store) => self.store_var(&store.address, &store.value),
-                Instruction::Load(load) => self.load(&load.address, &load.dest),
-                Instruction::Call(call) => self.call_func(call),
-                Instruction::Add(add) => {
-                    self.int_bin_operations(&add.operand0, &add.operand1, &add.dest, BinOps::Add)
-                }
-                Instruction::FAdd(fadd) => {
-                    self.fl_bin_operations(&fadd.operand0, &fadd.operand1, &fadd.dest, BinOps::Add)
-                }
-                Instruction::FPExt(fpext) => {
-                    self.fpext(&fpext.operand, &fpext.to_type, &fpext.dest)
-                }
-                Instruction::Sub(sub) => {
-                    self.int_bin_operations(&sub.operand0, &sub.operand1, &sub.dest, BinOps::Sub)
-                }
-                Instruction::Mul(mul) => {
-                    self.int_bin_operations(&mul.operand0, &mul.operand1, &mul.dest, BinOps::Mul)
-                }
-                Instruction::UDiv(udiv) => {
-                    self.int_bin_operations(&udiv.operand0, &udiv.operand1, &udiv.dest, BinOps::Div)
-                }
-                Instruction::SDiv(sdiv) => {
-                    self.int_bin_operations(&sdiv.operand0, &sdiv.operand1, &sdiv.dest, BinOps::Div)
-                }
-                Instruction::URem(urem) => {
-                    self.int_bin_operations(&urem.operand0, &urem.operand1, &urem.dest, BinOps::Rem)
-                }
-                Instruction::SRem(srem) => {
-                    self.int_bin_operations(&srem.operand0, &srem.operand1, &srem.dest, BinOps::Rem)
-                }
-                Instruction::FSub(fsub) => {
-                    self.fl_bin_operations(&fsub.operand0, &fsub.operand1, &fsub.dest, BinOps::Sub)
-                }
-                Instruction::SExt(sext) => self.sext(&sext.operand, &sext.to_type, &sext.dest),
-                Instruction::FMul(fmul) => {
-                    self.fl_bin_operations(&fmul.operand0, &fmul.operand1, &fmul.dest, BinOps::Mul)
+    fn it_bb(&mut self, func: &Function, bb_name: name::Name) {
+        for bb in func.basic_blocks.iter() {
```

Comment:
This is super inefficient, since in the worst case we need to iterate over all blocks in a function to find the next block to execute (and some functions can have thousands of blocks). `Function` has a `get_bb_by_name` method you can use.

Reply:
You're right. That's a bit surprising but can't be helped then. Probably still easier to use their function and leave add a `// PERF: ` comment that this is inefficient (which isn't great for an interpreter).

We can optimise this at a later stage. Can you think of a way to do that?

Reply:
Presumably a HashMap but this will require adding all name:block pairs to the HashMap will might take a lot of time if there are a lot of blocks.

Reply:
Yes. We have two choices:
1) Everytime you lookup a block add the result to a cache, so the next time you try to find it you can read it from the cache.
2) Precompute the indexes for each block right after loading in the module.

For an interpreter 2) is probably the better choice. It slows down startup a bit, but will make interpreter performance more stable. But we don't really need to worry about this just now. We can keep it in mind for a future optimisation.